network/ieee802_11_prism.ksy

```diff
@@ -0,0 +1,63 @@
+meta:
+  id: ieee802_11_prism
+  title: prism II headers for 802.11 capture
+  license: Unlicense
+  endian: le
+  encoding: ascii
+
+doc-ref:
+  - https://www.tcpdump.org/linktypes/LINKTYPE_IEEE802_11_PRISM.html
+
+seq:
+  - id: message_code
+    type: u4
+  - id: size
+    type: u4
+  - id: device_name
+    type: strz
+    size: 16
+  - id: fields
+    type: fields
+    size: fields_size
+  - id: body
+    size-eos: true
+instances:
+  header_size:
+    value: 4 + 4 + 16
+  fields_size:
+    value: size - header_size
+types:
+  fields:
+    seq:
+      - id: fields
+        type: field
+        repeat: eos
+    types:
+      field:
+        seq:
+          - id: flags
+            type: u1
+          - id: type
+            type: u1
+            enum: type
+          - id: unknown
+            type: u2
+            doc: usually zero
```

Comment:
According to Wireshark with the loaded [`80211_headers.pcap`](https://web.archive.org/web/20101116031635if_/http://www.raulsiles.com:80/downloads/80211_headers.pcap) file that you provide and also to the spec at https://www.tcpdump.org/linktypes/LINKTYPE_IEEE802_11_PRISM.html you specified in the `doc-ref`, this is not the correct interpretation.

These 4 bytes that I highlighted form together in both Wireshark and the mentioned specification a single `u4` integer, which is mapped to an enum (with integer values like `0x00020044` = MAC time etc.). In the [spec](https://www.tcpdump.org/linktypes/LINKTYPE_IEEE802_11_PRISM.html), there are two different enums presented, one for packets with the msgcode value 0x00000041 and the other for packets with the msgcode value 0x00000044. 

Of course, you don't have to stick to this approach, it's OK to split the single enum field into multiple attributes, but it has to parse the right values. The first packet in the file `80211_headers.pcap` gets parsed by Wireshark like this:

```yaml
Prism capture header
    Message Code: 0x00000044
    Message Length: 144
    Device Name: ath0
    DID Host Time -23742
        DID: Host Time (0x00010044)
        Status: Supplied (0)
        Length: 4
        Host Time: 4294943554
    DID Mac Time 186645954
        DID: Mac Time (0x00020044)
        Status: Supplied (0)
        Length: 4
        MAC timestamp (lower 32 bits): 186645954
    # ...
```

whereas your KSY reads this (data exported to JSON by the [Web IDE](https://ide.kaitai.io/devel/)):

```js
  "body": {
    "messageCode": 0x44,
    "size": 144,
    "deviceName": "ath0",
    "fields": {
      "fields": [
        {
          "flags": 0x44,
          "type": 0,
          "unknown": 1,
          "isSupplied": 0,
          "size": 4,
          "value": 4294943554
        },
        {
          "flags": 0x44,
          "type": 0,
          "unknown": 2,
          "isSupplied": 0,
          "size": 4,
          "value": 186645954
        },
        // ...
```

Note that the DID field type isn't recognized at all by your spec. 

I can see the motive of splitting this field the way you did it, but it unfortunately parses wrong values when the `message_code` is `0x44`. The reason is that though the enum values in case of `message_code == 0x44` appear to be similar as in case of `message_code == 0x41`, the "changing number" is at a different position. From the [specification](https://www.tcpdump.org/linktypes/LINKTYPE_IEEE802_11_PRISM.html):

```
For packets with the msgcode value 0x00000041, the DID field values are:
0x00001041 - host time;
0x00002041 - MAC time;
0x00003041 - channel;
...

For packets with the msgcode value 0x00000044, the DID field values are:
0x00010044 - host time;
0x00020044 - MAC time;
0x00030044 - channel;
...
```

So if you want to use the multi-attribute approach, one layout isn't enough and you'll probably have to be creative. What comes to my mind:

```yaml
types:
  field:
    seq:
      - id: type
        type: field_type
        size: 4
  field_type:
    seq:
      - id: code
        type: u1
        doc: seems to be equal to `_root.message_code`
      - id: skip
        type: b12
        if: code == 0x44
        doc: should be 0
      - id: type
        type: b4
        enum: type
enums:
  type:
    0x1: host_time
    0x2: mac_time
    0x3: channel
    # ...
```

Reply:
>These 4 bytes that I highlighted form together in both Wireshark and the mentioned specification a single u4 integer, which is mapped to an enum (with integer values like 0x00020044 = MAC time etc.).

Yes, I have initially done as in wireshark, but when I have read the docs in the net, it is clearly stated in them that the format is reverse engineered, so there is no correct public info on that. So I have splitted the field without any testing of the resulting spec.

I'm sorry for all the errors, I gonna try to fix them somewhen.

Reply:
It's OK, take your time, it's just that this PR is not yet finished. I'll mark it as a draft if you don't object, you can hopefully mark it as "ready for review" any time.

Reply:
Thanks. I have noticed that the second nibble (`1` or `4`) may mean the offset from the first nibble (`4`), so I have tried to interpret it as such.